integration-tests/src/mock_network/mod.rs

```diff
@@ -0,0 +1,287 @@
+};
+use near_primitives::block::GenesisId;
+use near_primitives::hash::CryptoHash;
+use near_primitives::types::{BlockHeight, ShardId};
+use std::collections::HashMap;
+use std::time::Duration;
+
+mod setup;
+
+/// MockPeerManagerActor mocks PeerManagerActor and responds to messages from ClientActor.
+/// Instead of sending these messages out to other peers, it simulates a network and reads
+/// the needed block and chunk content from storage.
+/// MockPeerManagerActor has the following responsibilities
+/// - Responds to BlockRequest, BlockHeadersRequest and PartialEncodedChunkRequest
+/// - Sends NetworkInfo to ClientActor periodically
+/// - Simulates block production and sends the most "recent" block to ClientActor
+struct MockPeerManagerActor {
+    /// Client address for the node that we are testing
+    client_addr: Recipient<NetworkClientMessages>,
+    /// Access a pre-generated chain history from storage
+    chain_history_access: ChainHistoryAccess,
+    /// Current network state for the simulated network
+    network_info: NetworkInfo,
+    /// Block production speed for the network that we are simulating
+    block_production_delay: Duration,
+}
+
+impl MockPeerManagerActor {
+    fn new(
+        client_addr: Recipient<NetworkClientMessages>,
+        genesis_config: &GenesisConfig,
+        chain: Chain,
+        peers_start_height: BlockHeight,
+        block_production_delay: Duration,
+    ) -> Self {
+        // for now, we only simulate one peer
+        // we will add more complicated network config in the future
+        let peer = FullPeerInfo {
+            peer_info: PeerInfo::random(),
+            chain_info: near_network_primitives::types::PeerChainInfoV2 {
+                genesis_id: GenesisId {
+                    chain_id: genesis_config.chain_id.clone(),
+                    hash: *chain.genesis().hash(),
+                },
+                height: peers_start_height,
+                tracked_shards: (0..genesis_config.shard_layout.num_shards()).collect(),
+                archival: false,
+            },
+            partial_edge_info: PartialEdgeInfo::default(),
+        };
+        let network_info = NetworkInfo {
+            connected_peers: vec![peer.clone()],
+            num_connected_peers: 1,
+            peer_max_count: 1,
+            highest_height_peers: vec![peer],
+            sent_bytes_per_sec: 0,
+            received_bytes_per_sec: 0,
+            known_producers: vec![],
+            peer_counter: 0,
+        };
+        Self {
+            client_addr,
+            chain_history_access: ChainHistoryAccess { chain },
+            network_info,
+            block_production_delay,
+        }
+    }
+
+    /// This function gets called periodically
+    /// When it is called, it increments peer heights by 1 and sends the block at that height
+    /// to ClientActor. In a way, it simulates peers that get upgraded and broadcast blocks
```

Comment:
what do you mean by 'upgraded'?

Reply:
yeah sorry for being misleading. I just meant nodes broadcasting new blocks